File: Maths/Medium/ncr.cpp
// { Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

 // } Driver Code Ends
// User function Template for C++

// https://practice.geeksforgeeks.org/problems/ncr1019/1/?difficulty[]=1&page=1&category[]=Mathematical&query=difficulty[]1page1category[]Mathematical#
// Given two integers n and r, find nCr. Since the answer may be very large, calculate the 
// answer modulo 109+7.
class Solution{
public:
    int nCr(int n, int r){
        int c[n+1][n+1];
        c[0][0] = 0;
        if(n < r) return 0;
        
        for(int i = 1; i <= n; i++)
        {
            c[i][0] = 1;
            c[i][i] = 1;
        }
        
        for(int i = 1; i <= n; i++)
        {
            for(int j = 1; j < i; j++)
                c[i][j] = (c[i-1][j-1] + c[i-1][j]) % 1000000007;
        }
        
        return c[n][r] % 1000000007;
    }
};

// { Driver Code Starts.

int main(){
    int t;
    cin>>t;
    while(t--){
        int n, r;
        cin>>n>>r;
        
        Solution ob;
        cout<<ob.nCr(n, r)<<endl;
    }
    return 0;
}  // } Driver Code Ends

File: Maths/Medium/ncr.py
#User function Template for python3
# https://practice.geeksforgeeks.org/problems/ncr1019/1/?difficulty[]=1&page=1&category[]=Mathematical&query=difficulty[]1page1category[]Mathematical#
# Given two integers n and r, find nCr. Since the answer may be very large, calculate the 
# answer modulo 109+7.
class Solution:
    def nCr(self, n, r):
        C = [[0 for x in range(r+1)] for x in range(n+1)]
        # Calculate value of Binomial
        # Coefficient in bottom up manner
        for i in range(n+1):
            for j in range(min(i, r)+1):
                # Base Cases
                if j == 0 or j == i:
                    C[i][j] = 1
     
                # Calculate value using
                # previously stored values
                else:
                    C[i][j] = (C[i-1][j-1] + C[i-1][j]) % 1000000007
     
        return (C[n][r]) % 1000000007


#{ 
#  Driver Code Starts
#Initial Template for Python 3

import sys
sys.setrecursionlimit(10**6)

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        n, r = [int(x) for x in input().split()]
        
        ob = Solution()
        print(ob.nCr(n, r))
# } Driver Code Ends